File: services/positus_service.py

import requests
import json
import logging
import os
from datetime import datetime
from services.oracle_service import conectar_oracle
from dotenv import load_dotenv
load_dotenv()

DEFAULT_TOKEN = os.environ.get("POSITUS_TOKEN")
DEFAULT_URL = os.environ.get("POSITUS_URL")
CIRURGIA_DOCUMENT_URL = (
    os.environ.get("POSITUS_CIRURGIA_DOCUMENT_URL")
    or os.environ.get("POSITUS_PREPARO_DOCUMENT_URL")
    or "https://cdn.positus.global/production/resources/samples/document.pdf"
)
CIRURGIA_DOCUMENT_FILENAME = (
    os.environ.get("POSITUS_CIRURGIA_DOCUMENT_FILENAME")
    or os.environ.get("POSITUS_PREPARO_DOCUMENT_FILENAME")
    or "document.pdf"
)
CIRURGIA_NAMESPACE = os.environ.get("POSITUS_CIRURGIA_TEMPLATE_NAMESPACE")
CIRURGIA_TEMPLATES = {
    "hospital_dia": {
        "name": os.environ.get("POSITUS_TEMPLATE_HOSPITAL_DIA") or "elpis_confirm_crur",
        "id": int(os.environ.get("POSITUS_TEMPLATE_HOSPITAL_DIA_ID") or 1),
        "document_url": (
            os.environ.get("POSITUS_CIRURGIA_HOSPITAL_DIA_DOCUMENT_URL")
            or "http://irssl.org.br/wp-content/uploads/2026/07/ORIENTACAO-PARA-PACIENTE-HOSPITAL-DIA.pdf"
        ),
        "document_filename": (
            os.environ.get("POSITUS_CIRURGIA_HOSPITAL_DIA_DOCUMENT_FILENAME")
            or "ORIENTACAO-PARA-PACIENTE-HOSPITAL-DIA.pdf"
        ),
    },
    "hospital_pernoite": {
        "name": os.environ.get("POSITUS_TEMPLATE_HOSPITAL_PERNOITE") or "elpis_cirur_hospital_pernoite",
        "id": int(os.environ.get("POSITUS_TEMPLATE_HOSPITAL_PERNOITE_ID") or 1),
        "document_url": (
            os.environ.get("POSITUS_CIRURGIA_HOSPITAL_PERNOITE_DOCUMENT_URL")
            or "http://irssl.org.br/wp-content/uploads/2026/07/ORIENTACOES-PARA-PACIENTES-PERNOITE.pdf"
        ),
        "document_filename": (
            os.environ.get("POSITUS_CIRURGIA_HOSPITAL_PERNOITE_DOCUMENT_FILENAME")
            or "ORIENTACOES-PARA-PACIENTES-PERNOITE.pdf"
        ),
    },
}

def _grava_log_envio(id_envio, id_empresa, id_template, numero_destino, status_http, response_text, payload_enviado, erro_texto=None):
    try:
        conn = conectar_oracle()
        cur = conn.cursor()
        cur.execute("""
            INSERT INTO DBAMV.tb_mensagem_log_envio
                (id_envio, id_empresa, id_template, numero_destino, status_http, response_text, payload_enviado, erro_texto, data_envio)
            VALUES (:1, :2, :3, :4, :5, :6, :7, :8, :9)
        """, (
            id_envio, id_empresa, id_template, numero_destino,
            status_http, response_text, json.dumps(payload_enviado, ensure_ascii=False),
            erro_texto, datetime.now()
        ))
        conn.commit()
        conn.close()
    except Exception as e:
        logging.error(f"⚠️ Falha ao gravar log de envio: {e}")

def enviar_confirmacao(numero, nome, especialidade, data, hora, unidade,
                       obs, token_empresa=None, url_empresa=None,
                       nome_template=None, namespace=None,
                       numero_whatsapp=None,
                       id_envio=None, id_empresa=None, id_template=None):

    token = token_empresa or DEFAULT_TOKEN
    url = url_empresa or DEFAULT_URL
    template_name = nome_template or "elpis_v1"
    template_namespace = namespace or "fac54ef7_5fcf_4efb_a331_3b02438f176c"

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }

    obs_formatted = obs and obs.strip() or "Sem observações."

    payload = {
        "from": numero_whatsapp or None,
        "to": numero,
        "type": "template",
        "template": {
            "namespace": template_namespace,
            "name": template_name,
            "language": {"policy": "deterministic", "code": "pt_BR"},
            "components": [{
                "type": "body",
                "parameters": [
                    {"type": "text", "text": nome or ""},
                    {"type": "text", "text": especialidade or ""},
                    {"type": "text", "text": data or ""},
                    {"type": "text", "text": hora or ""},
                    {"type": "text", "text": unidade or ""},
                    {"type": "text", "text": obs_formatted}
                ]
            }]
        },
        "messaging_product": "whatsapp"
    }

    # # SANDBOX TEST
    # payload_sandbox = {
    #     "from": numero_whatsapp or None,
    #     "to": numero,
    #     "type": "text",
    #     "text": {
    #         "body": "Oi, tudo bem? API SERVICE"
    #     }
    # }

    try:
        # # SANDBOX TEST
        # resp = requests.post(url, headers=headers, json=payload_sandbox, timeout=20) 
        resp = requests.post(url, headers=headers, json=payload, timeout=20)
        status_code = resp.status_code
        text = resp.text
        logging.info(f"Enviado para {numero}: {status_code} ({template_name})")

        _grava_log_envio(id_envio, id_empresa, id_template, numero, status_code, text, payload, None)
        return status_code, text

    except requests.RequestException as e:
        logging.error(f"Erro ao enviar para {numero}: {e}")
        _grava_log_envio(id_envio, id_empresa, id_template, numero, None, None, payload, str(e))
        return None, str(e)


def enviar_confirmacao_cirurgica(numero, nome, especialidade, data, hora, unidade,
                                 obs, tipo_internacao, token_empresa=None,
                                 url_empresa=None, namespace=None,
                                 numero_whatsapp=None, id_envio=None,
                                 id_empresa=None):
    template = CIRURGIA_TEMPLATES.get((tipo_internacao or "").strip().lower())
    if not template:
        return 400, "Selecione o tipo de internacao: hospital_dia ou hospital_pernoite."

    token = token_empresa or DEFAULT_TOKEN
    url = url_empresa or DEFAULT_URL
    template_name = template["name"]
    template_namespace = namespace or CIRURGIA_NAMESPACE
    obs_formatted = obs and obs.strip() or "Sem observacoes."

    payload_template = {
        "name": template_name,
        "language": {"policy": "deterministic", "code": "pt_BR"},
        "components": [
            {
                "type": "header",
                "parameters": [{
                    "type": "document",
                    "document": {
                        "link": template.get("document_url") or CIRURGIA_DOCUMENT_URL,
                        "filename": template.get("document_filename") or CIRURGIA_DOCUMENT_FILENAME
                    }
                }]
            },
            {
                "type": "body",
                "parameters": [
                    {"type": "text", "text": nome or ""},
                    {"type": "text", "text": especialidade or ""},
                    {"type": "text", "text": data or ""},
                    {"type": "text", "text": hora or ""},
                    {"type": "text", "text": unidade or ""},
                    {"type": "text", "text": obs_formatted}
                ]
            }
        ]
    }
    if template_namespace:
        payload_template["namespace"] = template_namespace

    payload = {
        "from": numero_whatsapp or None,
        "to": numero,
        "type": "template",
        "template": payload_template,
        "messaging_product": "whatsapp"
    }

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }

    try:
        resp = requests.post(url, headers=headers, json=payload, timeout=20)
        status_code = resp.status_code
        text = resp.text
        logging.info(f"Enviado para {numero}: {status_code} ({template_name})")
        _grava_log_envio(id_envio, id_empresa, template["id"], numero, status_code, text, payload, None)
        return status_code, text

    except requests.RequestException as e:
        logging.error(f"Erro ao enviar para {numero}: {e}")
        _grava_log_envio(id_envio, id_empresa, template["id"], numero, None, None, payload, str(e))
        return None, str(e)

File: src/services/positusService.js
const axios = require("axios");
const config = require("../config");
const logger = require("./logger");
const { getConnection } = require("./oracleService");

async function gravaLogEnvio({
  idEnvio,
  idEmpresa,
  idTemplate,
  numeroDestino,
  statusHttp,
  responseText,
  payloadEnviado,
  erroTexto
}) {
  let conn;
  try {
    conn = await getConnection();
    await conn.execute(
      `INSERT INTO DBAMV.tb_mensagem_log_envio
        (id_envio, id_empresa, id_template, numero_destino, status_http,
         response_text, payload_enviado, erro_texto, data_envio)
       VALUES (:idEnvio, :idEmpresa, :idTemplate, :numeroDestino, :statusHttp,
         :responseText, :payloadEnviado, :erroTexto, SYSDATE)`,
      {
        idEnvio,
        idEmpresa,
        idTemplate,
        numeroDestino,
        statusHttp,
        responseText,
        payloadEnviado: JSON.stringify(payloadEnviado),
        erroTexto
      },
      { autoCommit: true }
    );
  } catch (error) {
    logger.error("positus_log_insert_failed", { error, numeroDestino });
  } finally {
    if (conn) await conn.close();
  }
}

async function enviarTemplate({
  numero,
  payload,
  token,
  url,
  templateName,
  idEnvio,
  idEmpresa,
  idTemplate
}) {
  try {
    const response = await axios.post(url, payload, {
      timeout: Number(process.env.POSITUS_TIMEOUT_MS || 20000),
      proxy: false,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json"
      },
      validateStatus: () => true
    });

    const responseText =
      typeof response.data === "string" ? response.data : JSON.stringify(response.data);
    logger.info("positus_sent", { numero, status: response.status, templateName });
    await gravaLogEnvio({
      idEnvio,
      idEmpresa,
      idTemplate,
      numeroDestino: numero,
      statusHttp: response.status,
      responseText,
      payloadEnviado: payload,
      erroTexto: null
    });

    return [response.status, responseText];
  } catch (error) {
    logger.error("positus_send_failed", { error, numero, templateName });
    await gravaLogEnvio({
      idEnvio,
      idEmpresa,
      idTemplate,
      numeroDestino: numero,
      statusHttp: null,
      responseText: null,
      payloadEnviado: payload,
      erroTexto: error.message
    });
    return [null, error.message];
  }
}

async function enviarConfirmacao({
  numero,
  nome,
  especialidade,
  data,
  hora,
  unidade,
  obs,
  tokenEmpresa,
  urlEmpresa,
  nomeTemplate,
  namespace,
  numeroWhatsapp,
  idEnvio,
  idEmpresa,
  idTemplate = 1
}) {
  const token = tokenEmpresa || config.positus.token;
  const url = urlEmpresa || config.positus.url;
  const templateName = nomeTemplate || config.positus.defaultTemplateName;
  const templateNamespace = namespace || config.positus.defaultNamespace;
  const obsFormatted = obs?.trim() || "Sem observacoes.";

  const payload = {
    from: numeroWhatsapp || null,
    to: numero,
    type: "template",
    template: {
      namespace: templateNamespace,
      name: templateName,
      language: { policy: "deterministic", code: "pt_BR" },
      components: [
        {
          type: "body",
          parameters: [
            { type: "text", text: nome || "" },
            { type: "text", text: especialidade || "" },
            { type: "text", text: data || "" },
            { type: "text", text: hora || "" },
            { type: "text", text: unidade || "" },
            { type: "text", text: obsFormatted }
          ]
        }
      ]
    },
    messaging_product: "whatsapp"
  };

  return enviarTemplate({
    numero,
    payload,
    token,
    url,
    templateName,
    idEnvio,
    idEmpresa,
    idTemplate
  });
}

async function enviarPreparoCirurgico({
  numero,
  nome,
  data,
  unidade,
  tokenEmpresa,
  urlEmpresa,
  nomeTemplate,
  namespace,
  numeroWhatsapp,
  idEnvio,
  idEmpresa,
  idTemplate = config.positus.preparoTemplateId,
  documentUrl,
  documentFilename,
  textoPreparo
}) {
  const token = tokenEmpresa || config.positus.token;
  const url = urlEmpresa || config.positus.url;
  const templateName = nomeTemplate || config.positus.preparoTemplateName;
  const templateNamespace = namespace || config.positus.preparoNamespace;
  const preparoUrl = documentUrl || config.positus.preparoDocumentUrl;
  const preparoFilename = documentFilename || config.positus.preparoDocumentFilename;
  const bodyParameters = [
    { type: "text", text: nome || "Paciente" },
    { type: "text", text: data || "Data a confirmar" },
    { type: "text", text: unidade || "Unidade nao informada" }
  ];

  if (textoPreparo) {
    bodyParameters.push({ type: "text", text: String(textoPreparo).slice(0, 1000) });
  }

  if (!preparoUrl) {
    const message = "POSITUS_PREPARO_DOCUMENT_URL nao configurado para envio de preparo cirurgico.";
    logger.error("positus_preparo_not_configured", { numero, templateName, idEmpresa });
    return [null, message];
  }

  const payload = {
    from: numeroWhatsapp || null,
    to: numero,
    type: "template",
    template: {
      namespace: templateNamespace,
      name: templateName,
      language: { policy: "deterministic", code: "pt_BR" },
      components: [
        {
          type: "header",
          parameters: [
            {
              type: "document",
              document: {
                link: preparoUrl,
                filename: preparoFilename
              }
            }
          ]
        },
        {
          type: "body",
          parameters: bodyParameters
        }
      ]
    },
    messaging_product: "whatsapp"
  };

  return enviarTemplate({
    numero,
    payload,
    token,
    url,
    templateName,
    idEnvio,
    idEmpresa,
    idTemplate
  });
}

function resolveCirurgiaTemplate(tipoInternacao) {
  const key = String(tipoInternacao || "").trim().toLowerCase();
  const templates = config.positus.cirurgiaTemplates || {};
  return templates[key] || null;
}

function buildConfirmacaoButtons() {
  return [
    {
      type: "button",
      sub_type: "quick_reply",
      index: "0",
      parameters: [{ type: "payload", payload: "CONFIRMAR" }]
    },
    {
      type: "button",
      sub_type: "quick_reply",
      index: "1",
      parameters: [{ type: "payload", payload: "CANCELAR" }]
    }
  ];
}

async function enviarConfirmacaoCirurgica({
  numero,
  nome,
  especialidade,
  data,
  hora,
  unidade,
  obs,
  tipoInternacao,
  tokenEmpresa,
  urlEmpresa,
  namespace,
  numeroWhatsapp,
  idEnvio,
  idEmpresa,
  documentUrl,
  documentFilename
}) {
  const template = resolveCirurgiaTemplate(tipoInternacao);
  if (!template) {
    return [400, "Selecione um template de envio valido."];
  }

  const token = tokenEmpresa || config.positus.token;
  const url = urlEmpresa || config.positus.url;
  const templateName = template.name;
  const templateNamespace = namespace || config.positus.cirurgiaNamespace;
  const obsFormatted = obs?.trim() || "Sem observacoes.";
  const cirurgiaDocumentUrl = documentUrl || template.documentUrl || config.positus.cirurgiaDocumentUrl;
  const cirurgiaDocumentFilename =
    documentFilename || template.documentFilename || config.positus.cirurgiaDocumentFilename;

  if (!cirurgiaDocumentUrl) {
    const message = "POSITUS_CIRURGIA_DOCUMENT_URL nao configurado para envio de confirmacao cirurgica.";
    logger.error("positus_cirurgia_document_not_configured", { numero, templateName, idEmpresa });
    return [null, message];
  }

  const payload = {
    from: numeroWhatsapp || null,
    to: numero,
    type: "template",
    template: {
      name: templateName,
      language: { policy: "deterministic", code: template.languageCode || "pt_BR" },
      components: [
        {
          type: "header",
          parameters: [
            {
              type: "document",
              document: {
                link: cirurgiaDocumentUrl,
                filename: cirurgiaDocumentFilename
              }
            }
          ]
        },
        {
          type: "body",
          parameters: [
            { type: "text", text: nome || "" },
            { type: "text", text: especialidade || "" },
            { type: "text", text: data || "" },
            { type: "text", text: hora || "" },
            { type: "text", text: unidade || "" },
            { type: "text", text: obsFormatted }
          ]
        },
        ...(template.includeButtons === false ? [] : buildConfirmacaoButtons())
      ]
    },
    messaging_product: "whatsapp"
  };

  if (templateNamespace) {
    payload.template.namespace = templateNamespace;
  }

  logger.info("positus_cirurgia_payload_ready", {
    numero,
    templateName,
    tipoInternacao,
    documentUrl: cirurgiaDocumentUrl,
    documentFilename: cirurgiaDocumentFilename
  });

  return enviarTemplate({
    numero,
    payload,
    token,
    url,
    templateName,
    idEnvio,
    idEmpresa,
    idTemplate: template.id
  });
}

module.exports = { enviarConfirmacao, enviarPreparoCirurgico, enviarConfirmacaoCirurgica, gravaLogEnvio };
